### ISSUE TITLE
Solve Server-Side Templating Performance Issues

### ISSUE THREAD

**Issue by OnlyInAmerica** [FinalsClub/djKarma#49]
Currently, the browse pages takes over 3 seconds to render. We should investigate caching solutions.

Try opening two tabs:
1. Request the Browse Page
2. Request the Events page in another tab (It won't begin to load until after the browse page loads in the first tab)

You'll see that one request has the potential to block the site to others.

I've wanted to learn Memcached, and it's well documented in the [Django Docs](https://docs.djangoproject.com/en/dev/topics/cache/?from=olddocs/#memcached)

**Comment by sethwoodworth** [comment]
I think I have memcached running, but the internet is slow enough that I can't really tell a difference. Tomorrow I will set up a profiling tool to be able to test the numbers objectively.

**Comment by OnlyInAmerica** [comment]
Awesome.

**Comment by AndrewMagliozzi** [comment]
Sweet.

**Comment by sethwoodworth** [comment]
Without memcached running, the browse/schools page takes 1.43s to load just 

On first load (generating the cached version of the page), http://beta.karmanotes.org/browse/schools takes

```
1.24s waiting for the server
248ms receiving data
```

The time spent waiting on the server is the problem.

With django caching enabled and memcached running I get results like:

```
414ms waiting for the server
313ms receiving the page
```

On another request, I got:

```
213ms connecting
408ms waiting for the server
62ms receiving the data
```

The waiting for the server records are the problematic ones.

This is still a little iffy because the internet where I am has long load times, but I have tested this enough times that I can show a difference in load times between caching on and caching off, so I think this is taken care of for the moment.  There are still a few things that can be done to speed this up, but this will work for now.

**Comment by OnlyInAmerica** [comment]
A decent part of the rendering delay for these pages is the reverse foreign key lookups.
- how many notes are in this course? Go to notes table and filter all notes with fkey course=X
- how many courses are in this school? Go to course table and filter courses with fkey school=X

We can replace these with integer fields on each School / Course that are affected whenever a course / note is entered or deleted. We do this for site-wide school, course, and note counts. I'll do my best to get this in by Wednesday.

**Comment by sethwoodworth** [comment]
Make a new ticket for the query changes and reference this as the meta ticket.  When that one closes, we can close this one.

**Comment by sethwoodworth** [comment]
What do you think about using [django-debug-toolbar](https://github.com/django-debug-toolbar/django-debug-toolbar/) to help debug bad queries ?

**Comment by OnlyInAmerica** [comment]
That's a great idea. Perhaps we can dedicate a few days (maybe post-launch depending on how many tickets we can close ahead of that) for performance tuning.

**Comment by sethwoodworth** [comment]
Ok, no further changes are going to hit this before we go live.  Moving this to stretch, it's 'goodenough' for now.
